addon/content/reducer.js

```diff
@@ -181,6 +190,29 @@ function messages(state = initialMessages, action) {
       MessageUtils.setTags(action.msgUri, action.tags);
       return state;
     }
+    case "MSG_TOGGLE_TAG_BY_INDEX": {
+      // If we're here we've been passed the index of a tag to toggle on the
+      // given message as well as the existing tags.
+      const allTags = MailServices.tags.getAllTags({}).map(normalizeTag);
```

Comment:
We could probably put this part into a utility function within MessageUtils. Generally, I'd like us to be working towards making messages handled in the reducer with simple calls through to the backend. The idea being that at some stage in the future, we can replace the direct calls with a message passing boundary, which we're likely to need as we complete the move to WebExtensions.